replace pkg_resources everywhere

We use `pkg_resources` to pull templates out of the installed package but [pkg_resources is deprecated as an API](https://setuptools.pypa.io/en/latest/pkg_resources.html) and should be replaced with code based on [importlib.resources](https://docs.python.org/3.11/library/importlib.resources.html#module-importlib.resources).

These are the two occurences that are left:
* gromacs/config.py:from pkg_resources import resource_filename, resource_listdir
* tests/datafiles.py:from pkg_resources import resource_filename

3.9 - 3.11 runners fail with errors such as
```
==================================== ERRORS ====================================
______________ ERROR collecting tests/fileformats/test_convert.py ______________
tests/fileformats/test_convert.py:12: in <module>
    from gromacs.fileformats import convert
gromacs/__init__.py:216: in <module>
    from . import config
gromacs/config.py:328: in <module>
    templates = _generate_template_dict("templates")
gromacs/config.py:324: in _generate_template_dict
    for entry in resources.files(__name__).joinpath(dirname).iterdir()
../../../micromamba/envs/gromacs_wrapper/lib/python3.11/importlib/resources/_common.py:22: in files
    return from_package(get_package(package))
../../../micromamba/envs/gromacs_wrapper/lib/python3.11/importlib/resources/_common.py:55: in get_package
    raise TypeError(f'{package!r} is not a package')
E   TypeError: 'gromacs.config' is not a package
```

Python 3.12 works.

--- a/CHANGES
+++ b/CHANGES
@@ -3,7 +3,13 @@
 ==============================
 
 2024-??-??      0.9.1
+orbeckst
 
+* templates in `config.templates` are now stored as
+  `importlib.resources.abc.Traversable` (PosixPath-like) objects instead of
+  strings of paths to files in the file system because we switched from
+  pkg_resources to importlib.resources for managing access to included files
+  (#282)
 * fixed reporting of failures of GromacsCommand and DeprecationWarnings for use
   of \s in regular expression strings (#285)
 

--- a/gromacs/config.py
+++ b/gromacs/config.py
@@ -232,8 +232,7 @@ import subprocess
 import sys
 
 from configparser import ConfigParser
-
-from pkg_resources import resource_filename, resource_listdir
+from importlib import resources
 
 from . import utilities
 
@@ -307,25 +306,28 @@ path = [os.path.curdir, qscriptdir, templatesdir]
 
 
 def _generate_template_dict(dirname):
-    """Generate a list of included files *and* extract them to a temp space.
+    """Generate a list of included files *and* make them available.
 
-    Templates have to be extracted from the egg because they are used
-    by external code. All template filenames are stored in
-    :data:`config.templates`.
+    Template files are stored inside the package. They are made available as
+    :class:` importlib.resources.abc.Traversable` object with
+    :mod:`pathlib.Path`-like methods. They can normally be read with the usual
+    :func:`open` function but if a real file needs to be present, use
+    :func:`importlib.resources.as_file`.
+
+    All template entries are stored in :data:`config.templates`.
+
+    .. SeeAlso:: importlib.resources
+
     """
+    # for Python >= 3.12:
+    # replace `sys.modules[__name__].__package__` with `__name__`
     return dict(
-        (resource_basename(fn), resource_filename(__name__, dirname + "/" + fn))
-        for fn in resource_listdir(__name__, dirname)
-        if not fn.endswith("~")
+        (entry.name, entry)
+        for entry in resources.files(sys.modules[__name__].__package__)
+        .joinpath(dirname)
+        .iterdir()
+        if not entry.name.endswith(("~", "__pycache__", "__init__.py"))
     )
-
-
-def resource_basename(resource):
-    """Last component of a resource (which always uses '/' as sep)."""
-    if resource.endswith("/"):
-        resource = resource[:-1]
-    parts = resource.split("/")
-    return parts[-1]
 
 
 templates = _generate_template_dict("templates")
@@ -334,10 +336,10 @@ and queuing system scripts. They are provided as a convenience and
 examples but **WITHOUT ANY GUARANTEE FOR CORRECTNESS OR SUITABILITY FOR
 ANY PURPOSE**.
 
-All template filenames are stored in
-:data:`gromacs.config.templates`. Templates have to be extracted from
-the GromacsWrapper python egg file because they are used by external
-code: find the actual file locations from this variable.
+All templates are stored in :data:`gromacs.config.templates` as :class:`
+importlib.resources.abc.Traversable` objects with :mod:`pathlib.Path`-like
+methods. They can normally be read with the usual :func:`open` function but if
+a real file needs to be present, use :func:`importlib.resources.as_file`.
 
 **Gromacs mdp templates**
 
@@ -357,6 +359,7 @@ code: find the actual file locations from this variable.
    The queing system scripts are highly specific and you will need to add your
    own into :data:`gromacs.config.qscriptdir`.
    See :mod:`gromacs.qsub` for the format and how these files are processed.
+
 """
 
 #: The default template for SGE/PBS run scripts.

--- a/gromacs/fileformats/xvg.py
+++ b/gromacs/fileformats/xvg.py
@@ -605,7 +605,7 @@ class XVG(utilities.FileUtils):
 
     def _get_colors(self, color, columns):
         try:
-            cmap = matplotlib.cm.get_cmap(color)
+            cmap = matplotlib.colormaps[color]
             colors = cmap(
                 matplotlib.colors.Normalize()(
                     numpy.arange(len(columns[1:]), dtype=float)

--- a/gromacs/tools.py
+++ b/gromacs/tools.py
@@ -626,7 +626,7 @@ class Release(object):
     """
 
     gromacs_version = re.compile(
-        r"^G[rR][oO][mM][aA][cC][sS] version:" "\s*(VERSION)?\s*(?P<version>.+)$"
+        r"^G[rR][oO][mM][aA][cC][sS] version:\s*(VERSION)?\s*(?P<version>.+)$"
     )
 
     def __init__(self):

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -84,6 +84,13 @@ rmprefix = "release-"
 file = "gromacs/_version.py"
 
 
-
 [tool.black]
 extend-exclude = "tests/(test_core|fileformats/test_convert)\\.py"
+
+
+[tool.pytest.ini_options]
+filterwarnings = [
+	       "ignore::gromacs.AutoCorrectionWarning",
+	       "ignore::gromacs.LowAccuracyWarning",
+	       "ignore::numkit.LowAccuracyWarning",
+	       ]

--- a/tests/datafiles.py
+++ b/tests/datafiles.py
@@ -16,9 +16,14 @@ Basically, wheneever you need the path to the file, wrap the filename in ``dataf
 """
 
 
+import sys
 import os.path
-from pkg_resources import resource_filename
+from importlib import resources
 
 
 def datafile(name):
-    return resource_filename(__name__, os.path.join("data", name))
+    return (
+        resources.files(sys.modules[__name__].__package__)
+        .joinpath("data")
+        .joinpath(name)
+    )

--- a/tests/test_setup.py
+++ b/tests/test_setup.py
@@ -61,13 +61,17 @@ def energy_minimize(solvate, low_performance):
     TMPDIR, solvate_args = solvate
     nt = 2 if low_performance else 0
     with TMPDIR.as_cwd():
-        em_args = gromacs.setup.energy_minimize(
-            mdrun_args={"nt": nt},
-            integrator="steep",
-            emtol=5000,
-            maxwarn=1,
-            **solvate_args,
-        )
+        with pytest.warns(
+            gromacs.UsageWarning,
+            match="Unprocessed mdp option are interpreted as options for grompp",
+        ):
+            em_args = gromacs.setup.energy_minimize(
+                mdrun_args={"nt": nt},
+                integrator="steep",
+                emtol=5000,
+                maxwarn=1,
+                **solvate_args,
+            )
     return TMPDIR, em_args
 
 
@@ -106,18 +110,22 @@ def test_energy_minimize_custom_mdp(
     TMPDIR, solvate_args = solvate
     nt = 2 if low_performance else 0
     with TMPDIR.as_cwd():
-        try:
-            em_args = gromacs.setup.energy_minimize(
-                mdrun_args={"nt": nt}, mdp=mdp, emtol=5000, **solvate_args
-            )
-        except gromacs.exceptions.GromacsError as err:
-            # sometimes the em does not converge at all, e.g. 5.02988e+04 on atom 3277;
-            # (happens on Travis Linux with Gromacs 4.6.5 but not locally or on Travis OSX) so we
-            # re-run with a ridiculous tolerance so that we can at least test that the whole
-            # function can run to completion
-            em_args = gromacs.setup.energy_minimize(
-                mdrun_args={"nt": nt}, mdp=mdp, emtol=6e4, **solvate_args
-            )
+        with pytest.warns(
+            gromacs.UsageWarning,
+            match="Unprocessed mdp option are interpreted as options for grompp",
+        ):
+            try:
+                em_args = gromacs.setup.energy_minimize(
+                    mdrun_args={"nt": nt}, mdp=mdp, emtol=5000, **solvate_args
+                )
+            except gromacs.exceptions.GromacsError as err:
+                # sometimes the em does not converge at all, e.g. 5.02988e+04 on atom 3277;
+                # (happens on Travis Linux with Gromacs 4.6.5 but not locally or on Travis OSX) so we
+                # re-run with a ridiculous tolerance so that we can at least test that the whole
+                # function can run to completion
+                em_args = gromacs.setup.energy_minimize(
+                    mdrun_args={"nt": nt}, mdp=mdp, emtol=6e4, **solvate_args
+                )
     assert os.path.exists(em_args["struct"])
     assert os.path.exists(em_args["top"])
     assert em_args["mainselection"] == '"Protein"'

